File: app/api/v1/posts.py
"""Post API endpoints.

CRUD operations for campus buddy-finding posts with filtering and pagination.
"""

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user, get_db
from app.models.user import User
from app.schemas.common import APIResponse
from app.schemas.post import PostCreate, PostFilter, PostListResponse, PostResponse, PostUpdate
from app.services import post_service

router = APIRouter(prefix="/posts")


@router.post("", response_model=APIResponse[PostResponse], status_code=201)
async def create_post(
    data: PostCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> APIResponse[PostResponse]:
    """Create a new buddy-finding post.

    Posts include category, tags, target participant count, and optional
    expiry time for automatic closing.
    """
    post = await post_service.create_post(db, current_user.id, data)
    return APIResponse(code=201, message="Post created", data=PostResponse.model_validate(post))


@router.get("", response_model=APIResponse[PostListResponse])
async def list_posts(
    category: str | None = Query(default=None, description="Filter by category"),
    tag: str | None = Query(default=None, description="Filter by a single tag"),
    keyword: str | None = Query(default=None, description="Search title and description"),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
) -> APIResponse[PostListResponse]:
    """List active posts with optional filtering and pagination.

    Expired posts are always excluded. Results are sorted by newest first.
    """
    filters = PostFilter(
        category=category,
        tag=tag,
        keyword=keyword,
        page=page,
        page_size=page_size,
    )
    posts, total = await post_service.get_posts(db, filters)
    return APIResponse(
        data=PostListResponse(
            items=[PostResponse.model_validate(p) for p in posts],
            total=total,
            page=filters.page,
            page_size=filters.page_size,
        )
    )


@router.get("/{post_id}", response_model=APIResponse[PostResponse])
async def get_post(
    post_id: int,
    db: AsyncSession = Depends(get_db),
) -> APIResponse[PostResponse]:
    """Get a single post by ID. Lazily closes expired posts on access."""
    post = await post_service.get_post(db, post_id)
    return APIResponse(data=PostResponse.model_validate(post))


@router.put("/{post_id}", response_model=APIResponse[PostResponse])
async def update_post(
    post_id: int,
    data: PostUpdate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> APIResponse[PostResponse]:
    """Update a post. Only the author can edit.

    Only the fields provided in the request body are changed.
    """
    post = await post_service.update_post(db, post_id, current_user.id, data)
    return APIResponse(data=PostResponse.model_validate(post))


@router.delete("/{post_id}", response_model=APIResponse[PostResponse])
async def close_post(
    post_id: int,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> APIResponse[PostResponse]:
    """Close a post (soft delete). Only the author can close."""
    post = await post_service.close_post(db, post_id, current_user.id)
    return APIResponse(message="Post closed", data=PostResponse.model_validate(post))

File: frontend/src/stores/posts.ts
/**
 * Posts (邀约) store — post list, filtering, CRUD operations.
 * Placeholder: full implementation in Stage 2.
 */

import { defineStore } from 'pinia'
import { ref } from 'vue'
import type { PostResponse, PostListResponse } from '@/types'
import * as postsApi from '@/api/posts'

export const usePostsStore = defineStore('posts', () => {
  // ===== State =====
  const posts = ref<PostResponse[]>([])
  const currentPost = ref<PostResponse | null>(null)
  const total = ref(0)
  const page = ref(1)
  const pageSize = ref(20)
  const isLoading = ref(false)

  // ===== Filters =====
  const currentCategory = ref<string | undefined>(undefined)
  const currentTag = ref<string | undefined>(undefined)
  const currentKeyword = ref<string | undefined>(undefined)

  // ===== Actions =====

  async function fetchPosts() {
    isLoading.value = true
    try {
      const result: PostListResponse = await postsApi.listPosts({
        category: currentCategory.value,
        tag: currentTag.value,
        keyword: currentKeyword.value,
        page: page.value,
        page_size: pageSize.value,
      })
      posts.value = result.items
      total.value = result.total
    } finally {
      isLoading.value = false
    }
  }

  async function fetchPostById(id: number) {
    currentPost.value = await postsApi.getPost(id)
    return currentPost.value
  }

  async function createPost(data: Parameters<typeof postsApi.createPost>[0]) {
    const post = await postsApi.createPost(data)
    return post
  }

  async function updatePost(id: number, data: Parameters<typeof postsApi.updatePost>[1]) {
    const post = await postsApi.updatePost(id, data)
    if (currentPost.value?.id === id) currentPost.value = post
    return post
  }

  async function closePost(id: number) {
    await postsApi.closePost(id)
    if (currentPost.value?.id === id) {
      currentPost.value.status = 'closed'
    }
  }

  function setFilters(filters: { category?: string; tag?: string; keyword?: string }) {
    if (filters.category !== undefined) currentCategory.value = filters.category
    if (filters.tag !== undefined) currentTag.value = filters.tag
    if (filters.keyword !== undefined) currentKeyword.value = filters.keyword
    page.value = 1
  }

  function reset() {
    posts.value = []
    currentPost.value = null
    total.value = 0
    page.value = 1
  }

  return {
    posts,
    currentPost,
    total,
    page,
    pageSize,
    isLoading,
    currentCategory,
    currentTag,
    currentKeyword,
    fetchPosts,
    fetchPostById,
    createPost,
    updatePost,
    closePost,
    setFilters,
    reset,
  }
})
